# Classification and regression in Python with ApacheSpark

 - In this project I look at a real-world dataset and apply machine learning on it using Apache Spark. 

In [1]:
from IPython.display import Markdown, display
def printmd(string):
    display(Markdown('# <span style="color:red">'+string+'</span>'))


if ('sc' in locals() or 'sc' in globals()):
    printmd('<<<<<!!!!! It seems that you are running in a IBM Watson Studio Apache Spark Notebook. Please run it in an IBM Watson Studio Default Runtime (without Apache Spark) !!!!!>>>>>')


In [2]:
!pip install pyspark==2.4.5

     |████████████████████████████████| 217.8MB 54.2MB/s eta 0:00:01
     |████████████████████████████████| 204kB 45.6MB/s eta 0:00:01
  Stored in directory: /home/dsxuser/.cache/pip/wheels/bf/db/04/61d66a5939364e756eb1c1be4ec5bdce6e04047fc7929a3c3c
Successfully built pyspark


In [3]:
try:
    from pyspark import SparkContext, SparkConf
    from pyspark.sql import SparkSession
except ImportError as e:
    printmd('<<<<<!!!!! Please restart your kernel after installing Apache Spark !!!!!>>>>>')

In [4]:
sc = SparkContext.getOrCreate(SparkConf().setMaster("local[*]"))

spark = SparkSession \
    .builder \
    .getOrCreate()

- Let’s start by downloading the dataset and creating a dataframe. 

- This dataset can be found on DAX, the IBM Data Asset Exchange and can be downloaded for free.

- NOAA dataset: https://developer.ibm.com/exchanges/data/all/jfk-weather-data/

In [5]:
# delete files from previous runs
!rm -f jfk_weather*

# download the file containing the data in CSV format
!wget http://max-training-data.s3-api.us-geo.objectstorage.softlayer.net/noaa-weather/jfk_weather.tar.gz

# extract the data
!tar xvfz jfk_weather.tar.gz
    
# create a dataframe out of it by using the first row as field names and trying to infer a schema based on contents
df = spark.read.option("header", "true").option("inferSchema","true").csv('jfk_weather.csv')

# register a corresponding query table
df.createOrReplaceTempView('df')

--2020-06-06 16:33:15--  http://max-training-data.s3-api.us-geo.objectstorage.softlayer.net/noaa-weather/jfk_weather.tar.gz
Resolving max-training-data.s3-api.us-geo.objectstorage.softlayer.net (max-training-data.s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.196
Connecting to max-training-data.s3-api.us-geo.objectstorage.softlayer.net (max-training-data.s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.196|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2575759 (2.5M) [application/x-tar]
Saving to: ‘jfk_weather.tar.gz’

100%[======================================>] 2,575,759   4.07MB/s   in 0.6s   

2020-06-06 16:33:16 (4.07 MB/s) - ‘jfk_weather.tar.gz’ saved [2575759/2575759]

./._jfk_weather.csv
jfk_weather.csv


In [32]:
import pandas as pd
df = pd.read_csv("jfk_weather.csv")
df

/opt/conda/envs/Python36/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (8,10,11,12,13,14,15,16,17,20,22,23,25,40,63,64,67,68,69,84,85,86,87) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


STATION                     STATION_NAME  ELEVATION  LATITUDE  \
0       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
1       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
2       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
3       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
4       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
5       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
6       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
7       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
8       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
9       WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
10      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
11      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
12      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
13      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
14      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
15      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
16      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
17      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
18      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
19      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
20      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
21      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
22      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
23      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
24      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
25      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
26      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
27      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
28      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
29      WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
...            ...                              ...        ...       ...   
114515  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114516  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114517  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114518  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114519  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114520  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114521  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114522  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114523  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114524  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114525  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114526  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114527  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114528  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114529  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114530  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114531  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114532  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114533  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114534  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        3.4   40.6386   
114535  WBAN:94789  JFK INTERNATIONAL AIRPORT NY US        

- The dataset contains some null values, therefore schema inference didn’t work properly for all columns, in addition, a column contained trailing characters, so we need to clean up the data set first. 

In [6]:
import random
random.seed(42)

from pyspark.sql.functions import translate, col

df_cleaned = df \
    .withColumn("HOURLYWindSpeed", df.HOURLYWindSpeed.cast('double')) \
    .withColumn("HOURLYWindDirection", df.HOURLYWindDirection.cast('double')) \
    .withColumn("HOURLYStationPressure", translate(col("HOURLYStationPressure"), "s,", "")) \
    .withColumn("HOURLYPrecip", translate(col("HOURLYPrecip"), "s,", "")) \
    .withColumn("HOURLYRelativeHumidity", translate(col("HOURLYRelativeHumidity"), "*", "")) \
    .withColumn("HOURLYDRYBULBTEMPC", translate(col("HOURLYDRYBULBTEMPC"), "*", "")) \

df_cleaned =   df_cleaned \
                    .withColumn("HOURLYStationPressure", df_cleaned.HOURLYStationPressure.cast('double')) \
                    .withColumn("HOURLYPrecip", df_cleaned.HOURLYPrecip.cast('double')) \
                    .withColumn("HOURLYRelativeHumidity", df_cleaned.HOURLYRelativeHumidity.cast('double')) \
                    .withColumn("HOURLYDRYBULBTEMPC", df_cleaned.HOURLYDRYBULBTEMPC.cast('double')) \

df_filtered = df_cleaned.filter("""
    HOURLYWindSpeed <> 0
    and HOURLYWindSpeed IS NOT NULL
    and HOURLYWindDirection IS NOT NULL
    and HOURLYStationPressure IS NOT NULL
    and HOURLYPressureTendency IS NOT NULL
    and HOURLYPrecip IS NOT NULL
    and HOURLYRelativeHumidity IS NOT NULL
    and HOURLYDRYBULBTEMPC IS NOT NULL
""")

- Let's predict the value of one column based of some others. 
- It is sometimes helpful to print a correlation matrix. 

In [7]:
from pyspark.ml.feature import VectorAssembler
vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed","HOURLYWindDirection","HOURLYStationPressure"],
                                  outputCol="features")
df_pipeline = vectorAssembler.transform(df_filtered)
from pyspark.ml.stat import Correlation
Correlation.corr(df_pipeline,"features").head()[0].toArray()

array([[ 1.        ,  0.25478863, -0.26171147],
       [ 0.25478863,  1.        , -0.13377466],
       [-0.26171147, -0.13377466,  1.        ]])

- As it can be seen, HOURLYWindSpeed and HOURLYWindDirection correlate with 0.25478863 whereas HOURLYWindSpeed  and HOURLYStationPressure correlate with -0.26171147, this is a good sign if we want to predict HOURLYWindSpeed from HOURLYWindDirection and HOURLYStationPressure. 

- Note that the numbers can change over time as we are always working with the latest data.

- Since this is supervised learning, let’s split our data into train (80%) and test (20%) set.

In [18]:
from pyspark.ml.feature import VectorAssembler
vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed","HOURLYPressureTendency","HOURLYStationPressure"],
                                  outputCol="features")
df_pipeline = vectorAssembler.transform(df_filtered)
from pyspark.ml.stat import Correlation
Correlation.corr(df_pipeline,"features").head()[0].toArray()

array([[ 1.        , -0.01324305, -0.26171147],
       [-0.01324305,  1.        , -0.05821663],
       [-0.26171147, -0.05821663,  1.        ]])

- HOURLYWindSpeed and HOURLYPresseureTendency correlate with -0.01324305

In [19]:
splits = df_filtered.randomSplit([0.8, 0.2])
df_train = splits[0]
df_test = splits[1]

- Again, we can re-use our feature engineering pipeline.

In [20]:
from pyspark.ml.feature import StringIndexer, OneHotEncoder
from pyspark.ml.linalg import Vectors
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.feature import Normalizer
from pyspark.ml import Pipeline

vectorAssembler = VectorAssembler(inputCols=[
                                    "HOURLYWindDirection",
                                    "ELEVATION",
                                    "HOURLYStationPressure"],
                                  outputCol="features")

normalizer = Normalizer(inputCol="features", outputCol="features_norm", p=1.0)

- Now we define a function for evaluating our regression prediction performance.
- We’re using RMSE (Root Mean Squared Error) here , the smaller the better.

In [21]:
def regression_metrics(prediction):
    from pyspark.ml.evaluation import RegressionEvaluator
    evaluator = RegressionEvaluator(
    labelCol="HOURLYWindSpeed", predictionCol="prediction", metricName="rmse")
    rmse = evaluator.evaluate(prediction)
    print("RMSE on test data = %g" % rmse)

- Let’s run a linear regression model first for building a baseline.

In [22]:
# LR1

from pyspark.ml.regression import LinearRegression


lr = LinearRegression(labelCol="HOURLYWindSpeed", featuresCol='features', maxIter=100, regParam=0.0, elasticNetParam=0.0)
pipeline = Pipeline(stages=[vectorAssembler, normalizer,lr])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)
regression_metrics(prediction)

RMSE on test data = 5.29776


In [23]:
# LR1 with feature_norm

from pyspark.ml.regression import LinearRegression


lr = LinearRegression(labelCol="HOURLYWindSpeed", featuresCol='features_norm', maxIter=100, regParam=0.0, elasticNetParam=0.0)
pipeline = Pipeline(stages=[vectorAssembler, normalizer,lr])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)
regression_metrics(prediction)

RMSE on test data = 5.53438


- Now we’ll try a **Gradient Boosted Tree Regressor**

In [24]:
# GBT1

from pyspark.ml.regression import GBTRegressor
gbt = GBTRegressor(labelCol="HOURLYWindSpeed", maxIter=100)
pipeline = Pipeline(stages=[vectorAssembler, normalizer,gbt])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)
regression_metrics(prediction)

RMSE on test data = 5.08057


- Now let’s switch gears. 

- Previously, we tried to predict HOURLYWindSpeed, but now we predict HOURLYWindDirection. 

- In order to turn this into a **classification problem** we discretize the value using the **Bucketizer**. 

- The new feature is called HOURLYWindDirectionBucketized.

In [25]:
from pyspark.ml.feature import Bucketizer, OneHotEncoder
bucketizer = Bucketizer(splits=[ 0, 180, float('Inf') ],inputCol="HOURLYWindDirection", outputCol="HOURLYWindDirectionBucketized")
encoder = OneHotEncoder(inputCol="HOURLYWindDirectionBucketized", outputCol="HOURLYWindDirectionOHE")

- Again, we define a function in order to assess how we perform. 
- Here we just use the accuracy measure which gives us the fraction of correctly classified examples. 
- Again, 0 is bad, 1 is good.

In [26]:
def classification_metrics(prediction):
    from pyspark.ml.evaluation import MulticlassClassificationEvaluator
    mcEval = MulticlassClassificationEvaluator().setMetricName("accuracy") .setPredictionCol("prediction").setLabelCol("HOURLYWindDirectionBucketized")
    accuracy = mcEval.evaluate(prediction)
    print("Accuracy on test data = %g" % accuracy)

- Again, for baselining we use LogisticRegression.

In [27]:
# LGReg1

from pyspark.ml.classification import LogisticRegression
lr = LogisticRegression(labelCol="HOURLYWindDirectionBucketized", maxIter=10)
#,"ELEVATION","HOURLYStationPressure","HOURLYPressureTendency","HOURLYPrecip"

vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed","HOURLYDRYBULBTEMPC"],
                                  outputCol="features")

pipeline = Pipeline(stages=[bucketizer,vectorAssembler,normalizer,lr])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)
classification_metrics(prediction)

Accuracy on test data = 0.689552


- Let’s try some other Algorithms and see if model performance increases. 

- It’s also important to tweak other parameters like parameters of individual algorithms (e.g. number of trees for RandomForest) or parameters in the feature engineering pipeline, e.g. train/test split ratio, normalization, bucketing, etc.

In [28]:
# RF1

from pyspark.ml.classification import RandomForestClassifier
rf = RandomForestClassifier(labelCol="HOURLYWindDirectionBucketized", numTrees=30)

vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed","HOURLYDRYBULBTEMPC","ELEVATION","HOURLYStationPressure","HOURLYPressureTendency","HOURLYPrecip"],
                                  outputCol="features")

pipeline = Pipeline(stages=[bucketizer,vectorAssembler,normalizer,rf])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)
classification_metrics(prediction)

Accuracy on test data = 0.73318


In [29]:
# GBT2

from pyspark.ml.classification import GBTClassifier
gbt = GBTClassifier(labelCol="HOURLYWindDirectionBucketized", maxIter=100)

vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed","HOURLYDRYBULBTEMPC","ELEVATION","HOURLYStationPressure","HOURLYPressureTendency","HOURLYPrecip"],
                                  outputCol="features")

pipeline = Pipeline(stages=[bucketizer,vectorAssembler,normalizer,gbt])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)
classification_metrics(prediction)

Accuracy on test data = 0.736625


In [30]:
# RF1 with 10 trees

from pyspark.ml.classification import RandomForestClassifier
rf = RandomForestClassifier(labelCol="HOURLYWindDirectionBucketized", numTrees=10)

vectorAssembler = VectorAssembler(inputCols=["HOURLYWindSpeed","HOURLYDRYBULBTEMPC","ELEVATION","HOURLYStationPressure","HOURLYPressureTendency","HOURLYPrecip"],
                                  outputCol="features")

pipeline = Pipeline(stages=[bucketizer,vectorAssembler,normalizer,rf])
model = pipeline.fit(df_train)
prediction = model.transform(df_test)
classification_metrics(prediction)

Accuracy on test data = 0.732491
